hydro_cli/src/core/deployment.rs

```diff
@@ -5,10 +5,135 @@ use super::Service;
 use super::Host;
 
 use anyhow::Result;
+use futures::Future;
+use indicatif::MultiProgress;
 use tokio::sync::RwLock;
 
 use std::sync::Arc;
+use std::sync::Mutex;
+use std::sync::OnceLock;
 use std::sync::Weak;
+use std::time::Duration;
+
+static PROGRESS_TRACKER: OnceLock<Mutex<ProgressTracker>> = OnceLock::new();
+
+tokio::task_local! {
+    static CURRENT_GROUP: Vec<String>;
+}
+
+pub struct ProgressTracker {
+    multi_progress: MultiProgress,
+    current_count: usize,
+}
+
+impl ProgressTracker {
+    fn new() -> ProgressTracker {
+        ProgressTracker {
+            multi_progress: MultiProgress::new(),
+            current_count: 0,
+        }
+    }
+
+    pub fn start_task(&mut self, path: Vec<String>) -> Arc<indicatif::ProgressBar> {
+        self.current_count += 1;
+        let pb = Arc::new(self.multi_progress.add(indicatif::ProgressBar::new(100)));
+        pb.set_style(
+            indicatif::ProgressStyle::default_bar()
+                .template("{spinner:.green} {wide_msg} ({elapsed} elapsed)")
+                .unwrap(),
+        );
+        pb.set_message(path.join(" / "));
+        pb.enable_steady_tick(Duration::from_millis(100));
+        pb
+    }
+
+    pub fn start_task_progress(&mut self, path: Vec<String>) -> Arc<indicatif::ProgressBar> {
+        self.current_count += 1;
+        let pb = Arc::new(self.multi_progress.add(indicatif::ProgressBar::new(100)));
+        pb.set_style(
+            indicatif::ProgressStyle::default_bar()
+                .template("{spinner:.green} {wide_msg} {bar} ({elapsed} elapsed)")
+                .unwrap(),
+        );
+        pb.set_message(path.join(" / "));
+        pb.enable_steady_tick(Duration::from_millis(100));
+        pb
+    }
+
+    pub fn end_task(&mut self, pb: Arc<indicatif::ProgressBar>) {
+        self.current_count -= 1;
+        self.multi_progress.remove(&pb);
+
+        if self.current_count == 0 {
+            self.multi_progress.clear().unwrap();
+        }
+    }
+}
+
+impl ProgressTracker {
+    pub fn with_group<T, F: Future<Output = T>>(name: &str, f: F) -> impl Future<Output = T> {
+        let mut cur = CURRENT_GROUP
+            .try_with(|cur| cur.clone())
+            .unwrap_or_default();
+        cur.push(name.to_string());
+        CURRENT_GROUP.scope(cur, f)
+    }
+
+    pub fn leaf<T, F: Future<Output = T>>(name: String, f: F) -> impl Future<Output = T> {
+        let mut group = CURRENT_GROUP
+            .try_with(|cur| cur.clone())
+            .unwrap_or_default();
+        group.push(name);
```

Comment:
maybe this snippet could be a private helper fn too although ownership might break